File: lager1/uppsikt/submissions/wrong_answer/cheese_greedy_cover.py
#!/usr/bin/env python3
# @EXPECTED_GRADES@ WA WA WA WA WA
#
# Cheese: Greedy set cover on the visibility graph. Repeatedly lights the
# lighthouse covering the most uncovered positions. Finds a valid cover
# but not necessarily minimum, so the answer is >= optimal.

import sys
import heapq


def main():
    input = sys.stdin.readline
    n = int(input())
    p = list(map(int, input().split()))

    if n == 1:
        print(1)
        return

    # Build Cartesian tree using monotonic stack (max-heap)
    left = [-1] * n
    right = [-1] * n
    parent = [-1] * n
    stack = []

    for i in range(n):
        last = -1
        while stack and p[stack[-1]] < p[i]:
            last = stack.pop()
        if stack:
            right[stack[-1]] = i
            parent[i] = stack[-1]
        if last != -1:
            left[i] = last
            parent[last] = i
        stack.append(i)

    # Build adjacency list: tree edges + adjacent position edges
    adj = [set() for _ in range(n)]
    for i in range(n):
        if left[i] != -1:
            adj[i].add(left[i])
            adj[left[i]].add(i)
        if right[i] != -1:
            adj[i].add(right[i])
            adj[right[i]].add(i)
        if i + 1 < n:
            adj[i].add(i + 1)
            adj[i + 1].add(i)

    # Convert to frozen lists for fast iteration
    adj_list = [list(s) for s in adj]

    # Coverage set for each node: {itself} + neighbors
    # uncov_count[i] = number of uncovered nodes in coverage_set[i]
    covered = [False] * n
    uncov_count = [1 + len(adj_list[i]) for i in range(n)]

    # Max-heap using negative counts (Python min-heap)
    heap = [(-uncov_count[i], i) for i in range(n)]
    heapq.heapify(heap)

    ans = 0
    total_covered = 0

    while total_covered < n:
        # Pop until we find a valid entry
        while heap:
            neg_cnt, x = heapq.heappop(heap)
            if -neg_cnt == uncov_count[x] and uncov_count[x] > 0:
                break
        else:
            break

        # Light node x
        ans += 1

        # Find newly covered nodes in coverage_set[x]
        newly_covered = []
        if not covered[x]:
            newly_covered.append(x)
        for nb in adj_list[x]:
            if not covered[nb]:
                newly_covered.append(nb)

        # Mark them as covered
        for m in newly_covered:
            covered[m] = True
        total_covered += len(newly_covered)

        # Update uncov_count for affected nodes
        for m in newly_covered:
            # m is now covered, so for every node that has m in its coverage set,
            # decrement their uncov_count. m's coverage set neighbors are adj[m],
            # plus m itself.
            uncov_count[m] -= 1
            if uncov_count[m] >= 0:
                heapq.heappush(heap, (-uncov_count[m], m))
            for nb in adj_list[m]:
                uncov_count[nb] -= 1
                if uncov_count[nb] >= 0:
                    heapq.heappush(heap, (-uncov_count[nb], nb))

    print(ans)


if __name__ == "__main__":
    main()

File: lager1/uppsikt/submissions/wrong_answer/cheese_greedy_cover.cpp
// @EXPECTED_GRADES@ WA WA WA WA WA
//
// Cheese: Greedy set cover on the visibility graph. Repeatedly lights the
// lighthouse covering the most uncovered positions. Finds a valid cover
// but not necessarily minimum, so the answer is >= optimal.

#include <cstdio>
#include <vector>
#include <queue>
#include <set>
using namespace std;

int main() {
    int n;
    scanf("%d", &n);

    vector<int> p(n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &p[i]);
    }

    if (n == 1) {
        printf("1\n");
        return 0;
    }

    // Build Cartesian tree using monotonic stack (max-heap)
    vector<int> left(n, -1), right(n, -1), par(n, -1);
    vector<int> stk;

    for (int i = 0; i < n; i++) {
        int last = -1;
        while (!stk.empty() && p[stk.back()] < p[i]) {
            last = stk.back();
            stk.pop_back();
        }
        if (!stk.empty()) {
            right[stk.back()] = i;
            par[i] = stk.back();
        }
        if (last != -1) {
            left[i] = last;
            par[last] = i;
        }
        stk.push_back(i);
    }

    // Build adjacency list: tree edges + adjacent position edges
    // Use sets to deduplicate
    vector<set<int>> adj_set(n);
    for (int i = 0; i < n; i++) {
        if (left[i] != -1) {
            adj_set[i].insert(left[i]);
            adj_set[left[i]].insert(i);
        }
        if (right[i] != -1) {
            adj_set[i].insert(right[i]);
            adj_set[right[i]].insert(i);
        }
        if (i + 1 < n) {
            adj_set[i].insert(i + 1);
            adj_set[i + 1].insert(i);
        }
    }

    // Convert to vectors for fast iteration
    vector<vector<int>> adj(n);
    for (int i = 0; i < n; i++) {
        adj[i].assign(adj_set[i].begin(), adj_set[i].end());
    }
    adj_set.clear();

    // Coverage set for each node: {itself} + neighbors
    // uncov_count[i] = number of uncovered nodes in coverage_set[i]
    vector<bool> covered(n, false);
    vector<int> uncov_count(n);
    for (int i = 0; i < n; i++) {
        uncov_count[i] = 1 + (int)adj[i].size();
    }

    // Max-heap: (uncov_count, index)
    priority_queue<pair<int, int>> heap;
    for (int i = 0; i < n; i++) {
        heap.push({uncov_count[i], i});
    }

    int ans = 0;
    int total_covered = 0;

    while (total_covered < n) {
        // Pop until we find a valid entry
        int x = -1;
        while (!heap.empty()) {
            auto [cnt, idx] = heap.top();
            heap.pop();
            if (cnt == uncov_count[idx] && uncov_count[idx] > 0) {
                x = idx;
                break;
            }
        }
        if (x == -1) break;

        // Light node x
        ans++;

        // Find newly covered nodes in coverage_set[x]
        vector<int> newly_covered;
        if (!covered[x]) {
            newly_covered.push_back(x);
        }
        for (int nb : adj[x]) {
            if (!covered[nb]) {
                newly_covered.push_back(nb);
            }
        }

        // Mark them as covered
        for (int m : newly_covered) {
            covered[m] = true;
        }
        total_covered += (int)newly_covered.size();

        // Update uncov_count for affected nodes
        for (int m : newly_covered) {
            // m is now covered, so for every node that has m in its coverage set,
            // decrement their uncov_count. m's coverage set neighbors are adj[m],
            // plus m itself.
            uncov_count[m]--;
            if (uncov_count[m] >= 0) {
                heap.push({uncov_count[m], m});
            }
            for (int nb : adj[m]) {
                uncov_count[nb]--;
                if (uncov_count[nb] >= 0) {
                    heap.push({uncov_count[nb], nb});
                }
            }
        }
    }

    printf("%d\n", ans);
    return 0;
}
